File: blue-sdk-android/blue-sdk/src/main/kotlin/com/blue/sdk/manager/MedicationManager.kt
// MedicationManager.kt
// BlueSDK - 用药事件管理器（FR20~FR24）

package com.blue.sdk.manager

import com.blue.sdk.error.BlueError
import com.blue.sdk.internal.CommandQueue
import com.blue.sdk.model.MedicationRecord
import com.blue.sdk.enums.MedicationStatus
import com.blue.sdk.transport.CommandCode
import com.blue.sdk.transport.DPIDConstants
import com.blue.sdk.transport.FrameBuilder
import java.util.Calendar

internal class MedicationManager(private val commandQueue: CommandQueue) {

    fun sendMedicationNotification(status: Byte, completion: (Result<Unit>) -> Unit) {
        // 用药结果通知使用 DPID 0x6F（NOTIFICATION_OF_RESULTS）
        val data = byteArrayOf(DPIDConstants.NOTIFICATION_OF_RESULTS, 0x00, 0x00, 0x01, status)
        val frame = FrameBuilder.build(CommandCode.SEND_COMMAND, data)
        commandQueue.enqueue(CommandCode.SEND_COMMAND, frame) { result ->
            result.fold(
                onSuccess = { completion(Result.success(Unit)) },
                onFailure = { completion(Result.failure(it as BlueError)) }
            )
        }
    }

    companion object {
        fun parseMedicationEvent(data: ByteArray): Pair<Int, MedicationStatus>? {
            if (data.size < 11) return null
            val alarmIndex = DPIDConstants.alarmIndex(data[0]) ?: return null
            val status = MedicationStatus.fromByte(data[10]) ?: return null
            return Pair(alarmIndex, status)
        }

        /**
         * 解析用药记录上报帧（DPID=0x65）
         * 数据格式（15字节）：
         *   [0]=DPID(0x65) [1-3]=type/len(00 00 0B)
         *   [4]=闹钟DP点 [5-6]=年(高低) [7]=月 [8]=日
         *   [9]=闹钟小时 [10]=闹钟分钟 [11]=响铃小时 [12]=响铃分钟
         *   [13]=状态(01取药/02超时/03漏服/04提前) [14]=提前标志
         */
        fun parseMedicationRecord(data: ByteArray): MedicationRecord? {
            if (data.size < 14) return null

            val alarmDPID = data[4]
            val alarmIndex = DPIDConstants.alarmIndex(alarmDPID) ?: return null

            val yearHigh = data[5].toInt() and 0xFF
            val yearLow = data[6].toInt() and 0xFF
            val year = (yearHigh shl 8) or yearLow
            val month = data[7].toInt() and 0xFF
            val day = data[8].toInt() and 0xFF
            val alarmHour = data[9].toInt() and 0xFF
            val alarmMinute = data[10].toInt() and 0xFF
            val eventHour = data[11].toInt() and 0xFF
            val eventMinute = data[12].toInt() and 0xFF
            val statusByte = data[13]

            val status = MedicationStatus.fromByte(statusByte) ?: return null

            val cal = Calendar.getInstance().apply {
                set(year, month - 1, day, eventHour, eventMinute, 0)
                set(Calendar.MILLISECOND, 0)
            }
            // 使用帧中的日期+时分作为基础，加上当前秒和毫秒确保同一分钟内的记录不重复
            val baseTime = cal.timeInMillis
            val currentSecMs = System.currentTimeMillis() % 60000 // 当前秒+毫秒部分
            return MedicationRecord(
                timestamp = baseTime + currentSecMs,
                alarmIndex = alarmIndex,
                alarmHour = alarmHour,
                alarmMinute = alarmMinute,
                status = status
            )
        }
    }
}

File: blue-sdk-ios/BlueSDK/BlueSDK/Classes/Manager/MedicationManager.swift
// MedicationManager.swift
// BlueSDK - LX-PD02 智能药盒蓝牙通信 SDK
//
// 用药事件管理器：处理响铃、超时、用药结果、用药记录（FR20~FR24）

import Foundation

/// 用药事件管理器
final class MedicationManager {

    private let commandQueue: CommandQueue
    private let logger = BlueLogger.shared

    init(commandQueue: CommandQueue) {
        self.commandQueue = commandQueue
    }

    // MARK: - 下发用药结果通知（FR24）

    /// 向设备下发用药结果通知
    /// - Parameters:
    ///   - status: 通知状态（0=等待 1=响铃开始 2=错过 3=成功）
    ///   - completion: 结果回调
    func sendMedicationNotification(
        status: UInt8,
        completion: @escaping (Result<Void, BlueError>) -> Void
    ) {
        // 用药结果通知使用 0x6F（协议文档注释有歧义，以示例帧为准）
        let data: [UInt8] = [0x6F, 0x00, 0x00, 0x01, status]
        let frame = FrameBuilder.build(cmd: CommandCode.sendCommand, data: data)

        commandQueue.enqueue(cmd: CommandCode.sendCommand, frame: frame) { result in
            switch result {
            case .success:
                completion(.success(()))
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }

    // MARK: - 解析上报帧（FR20~FR23）

    /// 解析用药事件上报帧
    /// - Parameter data: 上报帧数据段（DPID=0x68）
    /// - Returns: 用药状态，解析失败返回 nil
    static func parseMedicationEvent(from data: [UInt8]) -> (alarmIndex: Int, status: MedicationStatus)? {
        // 数据格式：[DPID(0x68)][0x00][0x00][0x07][0x01][hour][minute][weekMask][advHigh][advLow][statusByte]
        // alarmIndex 直接从 DPID 推导（0x68 = alarm3 = index 3）
        guard data.count >= 11 else { return nil }
        let dpid = data[0]
        guard let alarmIndex = DPIDConstants.alarmIndex(for: dpid) else { return nil }
        let statusByte = data[10]
        guard let status = MedicationStatus.from(byte: statusByte) else { return nil }
        return (alarmIndex: alarmIndex, status: status)
    }

    /// 解析用药记录上报帧（DPID=0x65）
    /// 数据格式（15字节）：
    ///   [0]=DPID(0x65) [1-3]=type/len(00 00 0B)
    ///   [4]=闹钟DP点 [5-6]=年(高低) [7]=月 [8]=日
    ///   [9]=闹钟小时 [10]=闹钟分钟 [11]=响铃小时 [12]=响铃分钟
    ///   [13]=状态(01取药/02超时/03漏服/04提前) [14]=提前标志
    static func parseMedicationRecord(from data: [UInt8]) -> MedicationRecord? {
        guard data.count >= 14 else { return nil }

        let alarmDPID = data[4]
        guard let alarmIndex = DPIDConstants.alarmIndex(for: alarmDPID) else { return nil }

        let yearHigh = Int(data[5])
        let yearLow  = Int(data[6])
        let year     = (yearHigh << 8) | yearLow
        let month    = Int(data[7])
        let day      = Int(data[8])
        let alarmHour   = Int(data[9])
        let alarmMinute = Int(data[10])
        let eventHour   = Int(data[11])
        let eventMinute = Int(data[12])
        let statusByte  = data[13]

        guard let status = MedicationStatus.from(byte: statusByte) else { return nil }

        var components = DateComponents()
        components.year   = year
        components.month  = month
        components.day    = day
        components.hour   = eventHour
        components.minute = eventMinute
        let date = Calendar.current.date(from: components) ?? Date()

        return MedicationRecord(
            timestamp: Int64(date.timeIntervalSince1970 * 1000),
            alarmIndex: alarmIndex,
            alarmHour: alarmHour,
            alarmMinute: alarmMinute,
            status: status
        )
    }
}
